pkg/elle/core/depend.go

```diff
@@ -0,0 +1,212 @@
+package core
+
+import (
+	"fmt"
+	"log"
+	"strings"
+)
+
+// DataExplainer ...
+type DataExplainer interface {
+	// Given a pair of operations a and b, explains why b depends on a, in the
+	//    form of a data structure. Returns `nil` if b does not depend on a.
+	ExplainPairData(p1, p2 PathType) ExplainResult
+	// Given a pair of operations, and short names for them, explain why b
+	//  depends on a, as a string. `nil` indicates that b does not depend on a.
+	RenderExplanation(result ExplainResult, preName, postName string) string
+}
+
+// CombinedExplainer struct
+type CombinedExplainer struct {
+	Explainers []DataExplainer
+}
+
+// ExplainPairData find dependencies in a and b
+func (c *CombinedExplainer) ExplainPairData(p1, p2 PathType) ExplainResult {
+	panic("implement me")
+}
+
+// RenderExplanation render explanation result
+func (c *CombinedExplainer) RenderExplanation(result ExplainResult, preName, postName string) string {
+	return ""
+}
+
+// CombineExplainer combines explainers into one
+func CombineExplainer(explainers []DataExplainer) DataExplainer {
+	return &CombinedExplainer{explainers}
+}
+
+// Combine composes multiple analyzers
+func Combine(analyzers ...Analyzer) Analyzer {
+	panic("implement me")
+}
+
+type DependType string
+
+const (
+	RealtimeDepend  DependType = "realtime"
+	MonotonicDepend DependType = "monotonic"
+	ProcessDepend   DependType = "process"
+)
+
+type ExplainResult interface {
+	Type() DependType
+}
+
+type ProcessDependent struct {
+	Process int
+}
+
+func (_ ProcessDependent) Type() DependType {
+	return ProcessDepend
+}
+
+// ProcessExplainer ...
+type ProcessExplainer struct{}
+
+// ExplainPairData explain pair data
+func (e ProcessExplainer) ExplainPairData(p1, p2 PathType) ExplainResult {
+	if !p1.Process.Present() || !p2.Process.Present() {
+		return nil
+	}
+	if p1.Process.MustGet() == p2.Process.MustGet() && p1.Index < p2.Index {
+		return ProcessDependent{Process: p1.Process.MustGet()}
+	} else {
+		return nil
+	}
+}
+
+// RenderExplanation render explanation
+func (e ProcessExplainer) RenderExplanation(result ExplainResult, preName, postName string) string {
+	if result.Type() != ProcessDepend {
+		log.Fatalf("result type is not %s, type error", result.Type())
+	}
+	res := result.(ProcessDependent)
+	return fmt.Sprintf("process %d excuted %s before %s", res.Process, preName, postName)
+}
+
+// RealtimeExplainer is Realtime order explainer
+type RealtimeExplainer struct {
+	historyReference History
+	nextIndex        []int
+}
+
+// TODO: here it needs to get the preEnd and the postStart, it maybe complex and introduce another logic here.
+func (r RealtimeExplainer) ExplainPairData(preEnd, postEnd PathType) ExplainResult {
+	postStart := r.historyReference[r.nextIndex[postEnd.Index]]
```

Comment:
nextIndex only establishes a mapping from invoke to ok|fail|info from `RealtimeGraph`, we also need bi-direction here.